### ISSUE TITLE
Failure to compact due to maximum index size 64 GiB

### ISSUE THREAD

**Issue by thejosephstevens** [cortexproject/cortex#4705]
**Describe the bug**
Over the weekend we significantly expanded one of our clusters, pushing ~153M timeseries to our Cortex 1.11.1 cluster in a day. 

**To Reproduce**
1. Start Cortex 1.11.1 in single-tenant mode as microservices
2. Push ~153M timeseries to it
3. See the Compactor get to Stage 3 compaction and then fail with an error like this
```
level=error ts=2022-04-04T16:11:41.167458322Z caller=compactor.go:537 component=compactor msg="failed to compact user blocks" user=fake err="compaction: group 0@5679675083797525161: compact blocks [data/compact/0@5679675083797525161/01FZNK9ZNWHJWRXB2YNHZ9W8H9 data/compact/0@5679675083797525161/01FZPV9DZ1PGV0HHV9007P8ZWM]: \"data/compact/0@5679675083797525161/01FZTCHWRJHFN419FJPDM0MQNW.tmp-for-creation/index\" exceeding max size of 64GiB"
```
The two blocks referenced by this error are 12-hour blocks at level-3 compaction, each of which has an index of ~38 GB (summing together to ~76 GB > 64 GB).

**Expected behavior**
There should be a way of skipping this, forcing compaction, sharding, or something.

There's an upstream Thanos patch [here](https://github.com/thanos-io/thanos/pull/3409#issuecomment-721846904) which allows skipping compaction but it appears to not be used by Cortex today (found by @alvinlin123 in Cortex Slack). There's also a Thanos change [here](https://github.com/thanos-io/thanos/pull/3410/files) which would automatically skip past compaction if the block is too large.

Mimir appears to have the ability to get past this by sharding during compaction so multiple blocks are produced per day, each of which can have a smaller index.

**Environment:**
 - Infrastructure: Kubernetes (GKE)
 - Deployment tool: Jinja (derived from the cortex-jsonnet manifest)

**Storage Engine**
- [x] Blocks
- [ ] Chunks

**Additional Context**

**Comment by alvinlin123** [comment]
There is a related [Prometheus issue](https://github.com/prometheus/prometheus/issues/5868)

**Comment by alvinlin123** [comment]
After PR #4707  I still need to implement the part that auto skip compaction for humongous blocks

**Comment by alvinlin123** [comment]
Hmm looks like I can't just use Thanos `largeTotalIndexSizeFilter` for Cortex's `ShuffleShardingPlanner` because `ShuffleShardingPlanner` is coupled with non-exported `tsdbBasedPlanner` struct.

**Comment by alvinlin123** [comment]
Another relevant Thanos issue https://github.com/thanos-io/thanos/issues/3068 to track the sharding work

**Comment by thejosephstevens** [comment]
For what it's worth, we migrated to Mimir and no longer have issues with this. I think this still makes sense to implement in Cortex since it's a pretty brutal issue to hit (and effectively puts a hard cap on max tenant size), but it doesn't need to be kept open for us. If you'd like to close it for clean-up, feel free.

**Comment by jeromeinsf** [comment]
@thejosephstevens we are looking on how to address this in Cortex.

**Comment by yeya24** [comment]
@alvinlin123 I don't think we can close this issue since only the proposal gets merged. Please reopen it.

**Comment by alvinlin123** [comment]
@yeya24  you are right; closing this issue was a mistake. Thanks!

I still have to learn to pay attention when merging PR and not which issue may get incorrectly closed :)

**Comment by julienlau** [comment]
FYI : This is the default max value of this hidden flag of thanos :

```
	// TODO(bwplotka): This is short term fix for https://github.com/thanos-io/thanos/issues/1424, replace with vertical block sharding https://github.com/thanos-io/thanos/pull/3390.
	cmd.Flag("compact.block-max-index-size", "Maximum index size for the resulted block during any compaction. Note that"+
		"total size is approximated in worst case. If the block that would be resulted from compaction is estimated to exceed this number, biggest source"+
		"block is marked for no compaction (no-compact-mark.json is uploaded) which causes this block to be excluded from any compaction. "+
		"Default is due to https://github.com/thanos-io/thanos/issues/1424, but it's overall recommended to keeps block size to some reasonable size.").
		Hidden().Default("64GB").BytesVar(&cc.maxBlockIndexSize)
```

**Comment by alexqyle** [comment]
We have partitioning compaction implemented. This feature can avoid index size hard limit.

This feature is in https://github.com/cortexproject/cortex/releases/tag/v1.19.0

There is doc on how to use it: https://cortexmetrics.io/docs/guides/partition-compaction/
